src/lib/ast.rs

```diff
@@ -0,0 +1,623 @@
+            } else {
+                // Non-terminal node.
+                formatted.push_str(&format!("{}\n", node));
+                let first_child = node.first_child.unwrap();
+                let mut current_child = node.last_child.unwrap();
+                loop {
+                    stack.push(current_child);
+                    if current_child == first_child {
+                        break;
+                    }
+                    match self.nodes[current_child.index].previous_sibling {
+                        None => break,
+                        Some(id) => current_child = id,
+                    };
+                }
+            }
+        }
+        write!(f, "{}", formatted)
+    }
+}
+
+// #[derive(Clone, Debug)]
+/// An AST node within an `Arena`.
+pub struct Node<T, U> {
+    parent: Option<NodeId>,
+    previous_sibling: Option<NodeId>,
+    next_sibling: Option<NodeId>,
+    first_child: Option<NodeId>,
+    last_child: Option<NodeId>,
+    /// The actual data which will be stored within the tree.
+    pub data: T,
+    /// Type of this AST node, derived from grammar e.g. `expr`, `factor`, `term`.
+    pub ty: U,
+    /// Level of indentation, useful for pretty printing, formatting, etc.
+    pub indent: u16,
+}
+
+impl<T, U> Node<T, U> {
+    /// Create a new node, with data, but without a parent or children.
+    pub fn new(data: T, ty: U, indent: u16) -> Node<T, U> {
+        Node {
+            parent: None,
+            previous_sibling: None,
+            next_sibling: None,
+            first_child: None,
+            last_child: None,
+            data: data,
+            ty: ty,
+            indent: indent,
+        }
+    }
+
+    /// `true` if this node has no children.
+    pub fn is_leaf(&self) -> bool {
+        match self.first_child {
+            Some(_) => false,
+            None => true,
+        }
+    }
+
+    /// `true` if this node has no parent.
+    pub fn is_root(&self) -> bool {
+        match self.parent {
+            Some(_) => false,
+            None => true,
+        }
+    }
+
+    /// Return the ID of the parent node, if there is one.
+    pub fn parent(&self) -> Option<NodeId> {
+        self.parent
+    }
+
+    /// Return the ID of the first child of this node, if there is one.
+    pub fn first_child(&self) -> Option<NodeId> {
+        self.first_child
+    }
+
+    /// Return the ID of the last child of this node, if there is one.
+    pub fn last_child(&self) -> Option<NodeId> {
+        self.last_child
+    }
+
+    /// Return the ID of the previous sibling of this node, if there is one.
+    pub fn previous_sibling(&self) -> Option<NodeId> {
+        self.previous_sibling
+    }
+
+    /// Return the ID of the previous sibling of this node, if there is one.
+    pub fn next_sibling(&self) -> Option<NodeId> {
+        self.next_sibling
+    }
+}
+
+impl<T, U> fmt::Display for Node<T, U>
+    where T: fmt::Display,
```

Comment:
These say that `T` and `U` must already implement `fmt::Display` otherwise this trait won't work.